### ISSUE TITLE
Some Questions

### ISSUE THREAD

**Issue by ps915** [me-no-dev/KissKONG#2]
Hi, 
first of all thanks for your work. I have some Questions.

1. Is the TBS Unify SmartAudio supported?
2. Where can i see all features of this ODS?
3. Which OSDs are supported? PiggyOSD, MinimOSD,...
4. How to wire this OSD?

Thanks,
Phil

**Comment by Bandi017** [comment]
it works with any minim osd (KV Team MOD) as i know it

**Comment by me-no-dev** [comment]
Smart audio is not supported yet, simply because I do not have one. Protocol seems straight forward and should be a breeze to implement. Currently TrampHV is supported from the smart VTX.
Any MinimOSD hardware should work and if you want Tramp/SmartAudio support, you would need a minimODS with sensor pins broken out. I use the Current pin on MicroMinimOSD to interface with the Tramp VTX

**Comment by ps915** [comment]
ok, I'll ask Trappy for sending you the SmartAudio Protocol and maybe also a Unify for testing. 

Phil

**Comment by ps915** [comment]
SmartAudio is a protocol developed by TBS for OSD to VTx communication. All newer generation OSDs at TBS,
and all UNIFY PRO series VTX support SmartAudio V1.0, with the UNIFY PRO HV line we have launched
SmartAudio V2.0. The new version supports control for the new features (e.g. PitMode). SmartAudio is a
single-wire UART protocol, running over the Audio-wire. If you are a OSD or VTx developer interested in
adding support for SmartAudio, please create a support ticket with us so we can supply you with the technical
documentation:
● http://team-blacksheep.freshdesk.com/

**Comment by me-no-dev** [comment]
I looked at some code online. It's 4800-5100 baud single-line serial, like the Tramp is on 9600 and FrSky S-BUS and S-Port are on 57600. I took the Tramp for a spin today and it was horrid... :( my cheap no-brand VTX worked much-much better

**Comment by ps915** [comment]
i talked to trappy, the TBS boss. You can get a unify 4 free. Just order through the tbs website and dont pay. then send him the link from the order to his email: trappy[ät]gmail[dot]com

**Comment by me-no-dev** [comment]
I don't mind paying, but TBS is an US company, which means that if I order through their site, I will have to deal with the Customs and import fees, which in my country is a pain. If they can give me an online store in Europe that has them in stock, I'll get it from there.

**Comment by ps915** [comment]
they are based in Honkong China, maybe that's better for you? 
there are several shops in the, but no TBS store itself.

redbee.de, rctech.de n-facrory.de

**Comment by me-no-dev** [comment]
not available in either store :)

**Comment by me-no-dev** [comment]
well... same is true on TBS official site :D

**Comment by trackdork** [comment]
Any updates on this.  I am using this OSD to great effect and I also use the TBS Unify Pro.  If you've started coding I can pick up development if you are unable to... or test something.  I don't want to wait for the KISS FC support!

**Comment by me-no-dev** [comment]
I do not see it available, else I will get one and code the driver. I will use the same one for both tramp and smart audio. Code is pretty straight forward, but no device to test on yet.

**Comment by trackdork** [comment]
http://team-blacksheep.com/products/prod:unify_pro_sma ?

**Comment by me-no-dev** [comment]
Ordered ;)

**Comment by ps915** [comment]
awesome! =)

**Comment by ps915** [comment]
contact the Customer Support for the detaild protocol! =)

**Comment by trackdork** [comment]
it's already available here: https://github.com/betaflight/betaflight/files/479793/TBS.SmartAudio.Rev07.pdf

**Comment by trackdork** [comment]
So, I don't have access to an Arduino 've been messing around with a Particle Photon and my Unify and am able to send commands and config changes to it.  I'm going to try to refactor the code to fit in the KissKONG architecture and see if I can use your software serial implementation on it.  One thing I can report is that 90% of my packets are coming back with CRC errors, so you may want to plan for that in advance.  It takes me 3-10 tries to get back my initial settings request!

**Comment by me-no-dev** [comment]
I've read that baudrate can actually fluctuate between 4800 and 5200 bps. I have oscilloscope and logic analyzers and will have a look once I get the thing :)

**Comment by trackdork** [comment]
I had some time at lunch to implement an auto-baud detector which I will test when I get home.  My plan is to pick a middle value to start with (4800 gave me lots of grief, so maybe 4900), then periodically poll for status, immediately deactivating the serial device and using pulseIn to attempt to determine the baud rate.  That way we can automatically adjust on the fly.

**Comment by me-no-dev** [comment]
you can detect it based on the first 2 bytes (0xAA and 0x55). That is the reason they use those bytes

**Comment by me-no-dev** [comment]
so got my unify... it's all good, except there is no data on the audio port... manually settings stuff with the button works, but nothing happens when I'm using serial..

**Comment by trackdork** [comment]
I'm not sure what's going on.  I didn't have any issue accessing the data stream.  Do you have anything else connected to the audio port?  You said you had an oscilloscope, have you tried issuing a request and looking for a response?  There's no data stream that I can see until a request is issued.

**Comment by me-no-dev** [comment]
Yeah I already modified the soft serial code to support both tramp and tbs and can issue commands and all. I tried many things and the scope is connected all of the time. Audio goes through, data does not (I only tested the audio, not actually connected).

**Comment by trackdork** [comment]
that is really strange.  have you tried without audio connected at all?  There is an "app-note" of sorts discussing the need for capacitative coupling with specific R/C components when running audio and smart audio in parallel.  I have not experimented with that, I only ran tests with a single arduino I/O pin (3v logic/5v-tolerant) directly connected to the audio pin.  I started with this sketch: https://github.com/NightHawk32/SmartAudio-testing/blob/master/tbs_test.ino and saw success pretty much first try once I figured out the setup.
